rfc/0009-policy-report.md

```diff
@@ -0,0 +1,323 @@
+security scanning tools as Falco or Kubebench).
+
+Kubewarden will generate one `PolicyReport` object per each Namespace found inside of
+the cluster.
+Plus, it will maintain one single `ClusterPolicyReport` object to keep track
+of all the cluster-wide Kubernetes resources.
+
+## Namespace report
+
+A Namespace can be "affected" both by `ClusterAdmissionPolicy` and by `AdmissionPolicy` policies.
+When auditing a Namespace, only some Kubewarden policies are going to be interested
+about the events happening inside of that Namespace. In addition to that,
+the same Kubernetes resource could be evaluated by multiple policies. This will
+lead to several policy evaluations being performed.
+
+The `PolicyReport` is a namespaced CR that provides these information:
+
+* How many policy evaluations have been done during the audit run of the Namespace
+* How many policy evaluations led to a resource being accepted, rejected or
+  could not be performed because of an error.
+  Policy failures happen when an error occurs during the policy evaluation,
+  for example, because of a runtime panic
+* Detailed results about all the evaluations performed. An evaluation result contains
+  the following information: the policy being used, the resource that was
+  evaluated, the final outcome (accept/reject/failure) and the eventual message
+  returned by the policy
+
+The `PolicyReport` of the Namespace is kept updated during the lifetime of the
+cluster, meaning:
+
+* New Policies might become "interested" about the events happening inside of
+  the cluster. New evaluations results will be added to the PolicyReport
+* Existing Kubernetes objects and policies can change over time: some evaluations
+  could change their outcome (pass -> fail and vice versa). The evaluation results
+  must reflect that
+* Policies might be deleted/might be no longer "interested" about the Namespace:
+  the report must be updated to not container any reference the these policies
+  and their evaluation results
+* Kubernetes resources might be deleted: the report must be updated to not contain
+  any reference to them
+
+## Cluster report
+
+All the Kubernetes resources that are not namespaced, are going to be audited
+separately.
+The auditing results are going to be saved into a Custom Resource called:
+`ClusterPolicyReport`.
+
+Given `ClusterPolicyReport` has the same structure as `PolicyReport`, Kubewarden
+will handle it in the same way described in the previous section.
+
+## `PolicyReport` detailed overview
+
+For each Namespace, Kubewarden will find all the policies that are interested
+about actions happening inside of it. This is going to be a mix of
+some `ClusterAdmissionPolicy` policies and all the `AdmissionPolicy` policies
+that are defined inside of the Namespace.
+
+Next, this list of policies is going to be filtered: only policies interested
+in `CREATE` events are going to be considered.
+The audit report is not going to simulate `DELETE` and `UPDATE` events.
```

Comment:
Won't we miss important evaluations when using only CREATE policies?

I'm trying to image an scenario where an user have a policy to check important resource configuration to avoid service disruption if the resource misses some important info during an update (e.g. environment variable, container settings, etc). In this scenario, our audit report will not report issues and our users will not be notified upfront.

Reply:
If that policy checks for CREATE, UPDATE, then we will simulate creating that resource that is missing the environment variable, and all is fine. If it's only UPDATE, it will not get checked by the continous audit feature and it will not go into the  PolicyReport because of it.

But it will get picked by the usual behaviour of policies when the real UPDATE operation happens. And then, we need to decide if we want to also log real request outcomes in a separate PolicyReport.

Reply:
We need a predictable way to generate the payload used to simulate the event. This can be done only for `CREATE` and `DELETE` operations. `UPDATE` operations are hard to simulate, each policy has its own internal logic that can lead to different outcomes based on what we give as an input.

In the beginning I would start with just focusing on `CREATE` operations, later we might want to expand to cover also `DELETE` ops.